### PR TITLE
Re-enable automated CSRF Protection in Spring Security for Server Webapp

### PR DESCRIPTION
**Pull request by tdonohue**
## References
* This issue was discovered by LGTM, as (currently) our `WebSecurityConfiguration` includes `csrf().disabled()`.  See LGTM Rule: https://lgtm.com/rules/1511220334423/
* **MUST BE MERGED WITH ANGULAR PR:** https://github.com/DSpace/dspace-angular/pull/833  (If this PR is merged prior to the Angular PR, then authentication/authorization will no longer work in the Angular UI as every auth request will return a `403 Forbidden`)

## Description
As noted in that [LGTM rule](https://lgtm.com/rules/1511220334423/), Spring Security enables this CSRF protection by default and recommends it "for any request that could be processed by a browser client by normal users".

**Simply put, this PR reenables CSRF protection in Spring Security, and configures it based on [Angular best practices](https://angular.io/guide/http#security-xsrf-protection).**

Previously, we had chosen to *disable* this default CSRF protection because (as originally designed) we planned to make _all authentication checks_ via the `Authorization: Bearer...` Header.  If Authentication checks only occur via Headers, then CSRF protection is unnecessary (as [CSRF only applies to Cookies](https://security.stackexchange.com/questions/62080/is-csrf-possible-if-i-dont-even-use-cookies)).

However, we changed this behavior recently when implementing Shibboleth Authentication in PR #2651.  Shibboleth implementation required us to also support Cookie-based authentication checks.

On current `main`, while we _primarily_ still do authentication checks via that `Authorization` Header, we **now do support authentication checks via Cookie**.  See this newly added IT `testStatusPasswordAuthenticatedWithCookie()` which verifies that: https://github.com/DSpace/DSpace/commit/2d489ad70a1795551b20222ab8b4833607b14ac4

Because of this newly added behavior (and the fact that we plan to store additional session/user info in Cookies, see #2807), we really should have CSRF protection enabled.

**NOTE: After this is accepted, we will need to make minor updates to the REST Contract.**  I promise to make these changes once this has been accepted. 
1. All modification requests (e.g. POST, PUT, PATCH, DELETE) will require CSRF verification (X-XSRF-TOKEN header) to succeed. Otherwise a 403 Forbidden error will be returned.  Read only requests (GET, OPTIONS) do NOT require CSRF verification.
2. Our CSRF verification process aligns with the recommendations of Spring Security + Angular
    * The REST API sends a `XSRF-TOKEN` Cookie in each request.
    * When making a modification request, the Client **must** include a `X-XSRF-TOKEN` Header in the request, with the value of the `XSRF-TOKEN` Cookie in that Header.

## Instructions for Reviewers
Reviewers should:
1. Verify code & tests look good
2. Verify HAL Browser still works (it needed minor updates to support CSRF protection), especially authentication & access restricted endpoints.
3. (Optionally) test it with Shibboleth authentication as well. (However, I don't believe it should be affected.)

List of changes in this PR:
* Re-enabled CSRF protection in `WebSecurityConfiguration`
    * Took the opportunity to restructure `WebSecurityConfiguration` & add more inline comments, as it had gotten confusing.
* Allowed the `X-XSRF-TOKEN` header in `Application` settings
* Added CSRF protection support to all ITs.  See changes to `AbstractControllerIntegrationTest`. Required adding `spring-security-test` as a dependency
* Updated HAL Browser to support XSRF-TOKEN Cookie & X-XSRF-TOKEN Header
    * While making these modifications, I stumbled on a minor XSS security issue in HAL Browser described in https://github.com/mikekelly/hal-browser/pull/97.  So, I applied that fix.
* Minor updates to `AbstractBuilderCleanupUtil` to add in missing Builders.  I hit some strange random test errors that were solved by adding missing Builders into the cleanup list.
* **UPDATE:** This PR has been updated to ensure both the XSRF-TOKEN Cookie **and** the DSpace auth Cookie both have `SameSite=None; Secure`.  This ensures that both cookies can be used/sent cross domain (e.g. if UI is on a different host from backend, or if Shibboleth is on a different host from backend).

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:378]: NOTE for reviewers: This test is not directly related to CSRF protection.  However, I added it to prove that (depending on your `local.cfg` settings) the JWT Token (in header or Cookie) can be reused across IP addresses.  This is a real life scenario that could be encountered if a user is behind a proxy or ISP that switches IP addresses on them during their session.  That said, this sort of behavior _only safe_ if you have CSRF protection enabled.

Assuming we merge this CSRF protection PR, we could even consider removing entirely this `jwt.login.token.include.ip` option (in a follow-up PR).  It's not really necessary if CSRF protection is there, as you are protected even if your IP changes during your session.

**Comment by tdonohue** [comment]
@benbosman and @abollini : I've assigned this for your review (while I'm out on holiday).  Feel free to reassign to someone on your teams if you feel it can be delegated.  I'd just like to get more eyes on this while I'm out, that way we can either merge or discuss (if necessary) once I return.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java:405]: I'm in favor to remove the option. In such case we should rename the test `testReuseTokenWithDifferentIPWhenIPStored` in `testReuseTokenWithoutCSRFToken`

**Comment by KevinVdV** [comment]
@tdonohue The code looks good, HAL works just fine (can login & visit protected pages). But when I try to login using POSTMAN (so I can also test some other POST requests). I get a 403 & I don't even reach the PasswordAuthentication class, so I believe the XSRF-TOKEN is blocking me from authenticating.

Do I need to change something in my postman setup ?

**Comment by tdonohue** [comment]
@KevinVdV : I don't use Postman much, so I'm not an expert on how to set it up properly.  However, it sounds like Postman isn't handling this automatically.  

Essentially, the new *expected* behavior is...
1. Each request to the REST API now returns a `XSRF-TOKEN` Cookie
2. Anytime you want to do a non-GET request, you have to take the value of that `XSRF-TOKEN` Cookie and pass it back to the REST API in the `X-XSRF-TOKEN` header.

Googling around, this blog post appears to document how to get POSTMAN to do that behavior by default:
https://medium.com/@albertcito/test-your-api-with-postman-and-xsrf-token-59285107d5f7

If you could try that out, I'd appreciate it.  If it works, I can add a link to this resource in our REST Contract after this PR is merged...as I plan to make contract changes to better document the XSRF-TOKEN after this PR is accepted.

**Comment by KevinVdV** [comment]
@tdonohue Yeah postman already does some things with the XSRF-TOKEN cookie, but even taking the postman out of the equation, the following call from my CLI fails:

> curl --location --request POST 'http://localhost:8080/server/api/authn/login?user=*****&password=*******

This is my initial login when I don't have an "XSRF-TOKEN" cookie yet, this one should succeed no ?

**Comment by tdonohue** [comment]
@KevinVdV : No, it should not succeed as you've not included an X-XSRF-TOKEN header in your request. So, it's acting as it should.  Again, a reminder that CSRF protection is enabled for ANY non-GET (or HEAD) request.  So, any POST, PUT or DELETE *requires* that you pass along the XSRF-TOKEN in the X-XSRF-TOKEN header, **even for authentication requests**. 

This behavior is the default behavior recommended by Spring Security: https://docs.spring.io/spring-security/site/docs/3.2.0.CI-SNAPSHOT/reference/html/csrf.html

It's also the recommended behavior from OWASP in protecting against CSRF attacks: https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html

So, what you are seeing from the CLI is expected behavior.  We will need to update the REST Docs to make it clear that you must first make a GET/HEAD request to obtain an XSRF-TOKEN cookie, only then can you POST a valid login request (passing along the X-XSRF-TOKEN header).

**Review by KevinVdV — APPROVED** [review]
@tdonohue You are correct, I needed to the add the X-XSRF-TOKEN header (I passed the cookie along but forgot about the http header), once I did that I could login again & do other posts. 

Approving this one, thanks @tdonohue !

**Comment by tdonohue** [comment]
After some detailed analysis over the last few days, I've realized that the default Spring Security CSRF Protection **will not work for our needs as it relies too heavily on Cookies.**

* (As documented in the description above), it *requires* any Client to be able to read the `XSRF-TOKEN` Cookie and then *respond* with the same token in an `X-XSRF-TOKEN` Header.
* While this all works great if you are running the Angular UI and REST API _on the same domain or even on two parallel subdomains_, it **will not work if they run on entirely separate domains** (like our [demo UI](https://dspace7-demo.atmire.com/) and [demo REST API](https://dspace7.4science.cloud/server/)).  It also won't work if the REST API needs to speak to any client on a different domain (which seems like a use case we'd want to support).
* The reason this doesn't work when running on separate domains is a very simple one... Cookies don't work across domains (for security reasons, these are considered third-party cookies and are usually/often blocked by the browser).

**NEW APPROACH:** Taking a step back here, I've been starting to rethink this effort along these lines:

1. We only need CSRF protection when storing our Auth token in a Cookie. By default, we only store the Auth token in the `Authorization` Header, [which isn't susceptible to CSRF attacks](https://security.stackexchange.com/a/170414)
2. We only need our Auth token in a Cookie when Shibboleth is enabled (auth cookies were not even added until Shibboleth support was added in https://github.com/DSpace/DSpace/pull/2651)
3. Therefore, sites *without* Shibboleth should have an easy way to ensure the Auth token is *never stored in or read from* a Cookie. (Currently, this doesn't entirely exist, as our Auth system "falls back" to looking for a Cookie if the Header isn't found)
4. Additionally, sites *with* Shibboleth enabled need to have mitigation/protection against CSRF attacks on the required Auth Cookie.
    * We must ensure the Auth Cookie uses SameSite, is secure (HTTPS only) and disables usage by Javascript (HttpOnly).  This simply tightens the security of the cookie itself.
    * We should verify the Origin of all requests using the Auth Cookie. This will ensure the Auth Cookie cannot be passed to/from a different Origin: https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#verifying-origin-with-standard-headers
    * If, at some point in the future, we are able to find a way to _change our Shibboleth implementation to use the `Authorization` Header, then we could completely remove Auth Cookie support.

So, all the above being said. I'm going to close this PR as `won't fix` as I cannot find a way for this approach to work when the UI and REST API are on entirely separate domains. It also wouldn't work if the REST API needs to allow login access to _any client/system_ on a separate domain (which would limit third-party integrations with the REST API).

Instead, I will follow-up with a new PR which attempts the new approach to more tightly limit when Auth cookies are used *and* to secure them as much as we reasonably can.   In the meantime, feedback on the new approach is welcome.

**Pull request closed without merging**
